### PR TITLE
gui: Disable unavailable context menu items in transactions tab

### LINKED ISSUE #9192: Context menu options that are not available should be disabled (in transactions tab)
### Describe the issue
In Bitcoin QT, under the Transactions tab when right clicking on the transaction the context menu shows more option then it can always handle.

### Can you reliably reproduce the issue?
Yes.

#### If so, please list the steps to reproduce below:  
0. Open QT.  
1. Make an incoming transaction without label.  
2. Go to the Transactions tab.  
3. Rigth click on a transaction and select "Copy label".  

### Actual behaviour
The clipboard will be empty.  

### Expected behaviour
Copies the label. (What does not exist, of course.)

### What version of bitcoin-core are you using?
v0.13.1, Ubuntu PPA.  

### Any extra information that might be useful in the debugging process.
This behaviour is more problematic in edge cases.  
Like when you send funds with JoinMarket. In this case for example the "Copy address" option does not work.  

### Suggestion
I think when something cannot be copied that should not be present in the context menu in the first place.

### PR DESCRIPTION
**Pull request by kristapsk**
Fixes #9192.

### CODE REVIEW

**Comment by hebasto** [comment]
@kristapsk screenshots "before" and "after" are welcome for GUI-related PRs ;)

**Comment by kristapsk** [comment]
@hebasto Functionality seemed so simple and obvious (just menu items being disabled) that there isn't need for screenshots. But could add. All four pairs of possible before and after screenshots?

File: src/qt/transactionview.cpp [hunk 1]
```diff
@@ -395,6 +396,8 @@ void TransactionView::contextualMenu(const QPoint &point)
     hash.SetHex(selection.at(0).data(TransactionTableModel::TxHashRole).toString().toStdString());
     abandonAction->setEnabled(model->wallet().transactionCanBeAbandoned(hash));
     bumpFeeAction->setEnabled(model->wallet().transactionCanBeBumped(hash));
+    copyAddressAction->setEnabled(GUIUtil::hasEntryData(transactionView, 0, TransactionTableModel::AddressRole));
```
Comment (hebasto) [inline src/qt/transactionview.cpp]: How to test this line of the code?

Comment (kristapsk) [reply src/qt/transactionview.cpp]: I tested with some JoinMarket transactions, where I had partially imported JM wallet addresses as watchonly (some of inputs belonged to the wallet, but no output addresses).

**Comment by jonasschnelli** [comment]
Nice!
Concept ACK.

File: src/qt/guiutil.cpp [hunk 2]
```diff
@@ -250,6 +250,15 @@ QList<QModelIndex> getEntryData(QAbstractItemView *view, int column)
     return view->selectionModel()->selectedRows(column);
 }
 
+bool hasEntryData(QAbstractItemView *view, int column, int role)
+{
+    QModelIndexList selection = getEntryData(view, column);
+    if (!selection.isEmpty()) {
```
Comment (promag) [inline src/qt/guiutil.cpp]: nit, 
```
if (selection.isEmpty()) return false
return !selection.at(0).data(role).toString().isEmpty();
```

File: src/qt/transactionview.h [hunk 3]
```diff
@@ -77,6 +77,8 @@ class TransactionView : public QWidget
     QDateTimeEdit *dateTo;
     QAction *abandonAction;
     QAction *bumpFeeAction;
+    QAction *copyAddressAction;
```
Comment (promag) [inline src/qt/transactionview.h]: prefer `QAction* copyAddressAction{nullptr};` here and below and drop changes to constructor initializer list.

File: src/qt/guiutil.h [hunk 4]
```diff
@@ -77,6 +77,14 @@ namespace GUIUtil
      */
     QList<QModelIndex> getEntryData(QAbstractItemView *view, int column);
 
+    /** Check whether a field of the currently selected entry of a view is not empty. Does nothing if
+        nothing is selected.
+       @param[in] column  Data column to extract from the model
+       @param[in] role    Data role to extract from the model
+       @see  TransactionView::contextualMenu
+     */
+    bool hasEntryData(QAbstractItemView *view, int column, int role);
```
Comment (promag) [inline src/qt/guiutil.h]: Looks like this should be `const` but `getEntryData` is not. Would that be a big change?

**Review by promag — COMMENTED** [review]
Concept ACK.

**Comment by kristapsk** [comment]
Looks like I fucked this up. Any hints how to undo this git merge mess?

**Comment by kristapsk** [comment]
Ok, figured out myself, addressed comments by @promag.

**Review by nopara73 — APPROVED** [review]
Untested concept ACK, my original suggestion was the removal of non-functional context menu entries, but disabling is better, as it's used already:

image

**Comment by promag** [comment]
Concept ACK.

File: src/qt/guiutil.h [hunk 5]
```diff
@@ -68,14 +68,22 @@ namespace GUIUtil
...
     /** Return a field of the currently selected entry as a QString. Does nothing if nothing
         is selected.
        @param[in] column  Data column to extract from the model
        @see  TransactionView::copyLabel, TransactionView::copyAmount, TransactionView::copyAddress
      */
-    QList<QModelIndex> getEntryData(QAbstractItemView *view, int column);
+    QList<QModelIndex> getEntryData(const QAbstractItemView *view, int column);
+
+    /** Check whether a field of the currently selected entry of a view is not empty. Does nothing if
+        nothing is selected.
```
Comment (jonatack) [inline src/qt/guiutil.h:81]: This doc is for a boolean, so remove "Does nothing if nothing is selected."

Suggested change (or something similar/simpler):
```diff
-    /** Check whether a field of the currently selected entry of a view is not empty. Does nothing if
-        nothing is selected.
+    /** Returns true if the specified field of the currently selected view entry is not empty.
```

**Review by jonatack — COMMENTED** [review]
Thanks, good to see user-facing improvements like this.

Tested ACK f780d5a588075efdcd6d6fd097b251190c8d5e7e modulo docstring fix.

**Comment by jonatack** [comment]
Re-ACK 2b18fd2242a5899

`git diff f780d5a 2b18fd2` shows the only change since the previous review is the bool docstring fixup
```diff
-/** Check whether a field of the currently selected entry of a view is not empty. Does nothing if
-    nothing is selected.
+/** Returns true if the specified field of the currently selected view entry is not empty.
```

**Comment by kristapsk** [comment]
Is there anything anyone expects me to do more here or this is just forgotten in a long list of pending pull requests?

**Comment by jonatack** [comment]
@kristapsk the latter, I think. I agree this would be good, for now it needs a couple more reviewers' eyes on it.

**Comment by hebasto** [comment]
Concept ACK.

May I suggest to:
- split out refactor-only changes into a separate commit
- use [standard](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#contributor-workflow) prefixes in commit messages, e.g., `refactor: ...` and `qt: ...`
- apply [`clang-format-diff.py`](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy) for each commit

?

**Comment by jonasschnelli** [comment]
codereview utACK 2b18fd2242a589988fbb68205dae4afa0b8b3d34

**Pull request merged**